contracts/buy/PartyBuy.sol

```diff
@@ -0,0 +1,180 @@
+        _contribute();
+    }
+
+    // ======== External: Buy =========
+
+    /**
+     * @notice Buy the token by calling targetContract with calldata supplying value
+     * @dev Emits a Bought event upon success; reverts otherwise. callable by anyone
+     */
+    function buy(uint256 _value, address _targetContract, bytes calldata _calldata) external nonReentrant {
+        require(
+            partyStatus == PartyStatus.ACTIVE,
+            "PartyBuy::buy: party not active"
+        );
+        // check that value is not more than the maximum price set at deploy time
+        require(_value <= maxPrice, "PartyBuy::buy: can't spend over max price");
+        // check that value is not more than
+        // the maximum amount the party can spend while paying ETH fee
+        require(_value <= getMaximumSpend(), "PartyBuy::buy: insuffucient funds to buy token plus fee");
+        // execute the calldata on the target contract
+        address(_targetContract).call{value: _value}(_calldata);
+        // NOTE: we don't are if the call succeeded
+        // as long as the NFT is owned by the Party
+        require(_getOwner() == address(this), "PartyBuy::buy: failed to buy token");
+        // set partyStatus to WON
+        partyStatus = PartyStatus.WON;
+        // record totalSpent,
+        // send ETH fees to PartyDAO,
+        // fractionalize the Token
+        // send Token fees to PartyDAO & split proceeds to split recipient
+        uint256 _ethFee = _closeSuccessfulParty(_value);
+        // emit Bought event
+        emit Bought(msg.sender, _targetContract, _value, _ethFee, totalContributedToParty);
+    }
+
+    // ======== External: Fail =========
+
+    /**
+     * @notice If the token couldn't be successfully bought
+      * within the specified period of time, move to FAILED state
+      * so users can reclaim their funds.
+     * @dev Emits a Closed event upon finishing; reverts otherwise.
+     * callable by anyone after expiresAt
+     */
+    function closeParty() external {
+        require(
+            partyStatus == PartyStatus.ACTIVE,
+            "PartyBuy::closeParty: party not active"
+        );
+        require(expiresAt <= block.timestamp, "PartyBuy::closeParty: party has not timed out");
+        require(_getOwner() != address(this), "PartyBuy::closeParty: contract owns token");
+        // set partyStatus to LOST
+        partyStatus = PartyStatus.LOST;
+        // emit Closed event
+        emit Closed();
+    }
+
+    // ============ Internal ============
+
+    /**
+    * @notice Calculate ETH fee for PartyDAO
+    * NOTE: Remove this fee causes a critical vulnerability
+    * allowing anyone to exploit a PartyBuy via price manipulation.
+    * See Security Review in README for more info.
+    * @return _maxContributions the maximum amount that can be contributed to the party
+    */
+    function getMaximumContributions() public view returns (uint256 _maxContributions) {
+        uint256 _price = maxPrice;
```

Comment:
nit but why assign this to a variable?

Reply:
reading from storage is expensive, and we read the same variable twice in this function, so it is cheaper to load it from storage once